From acc9965a22977996069b36786c34c2f674ead800 Mon Sep 17 00:00:00 2001
From: Raj Kumar Bhagat <quic_rajkbhag@quicinc.com>
Date: Wed, 5 Jul 2023 15:33:15 +0530
Subject: [PATCH 2/2] ath12k: apply real noise floor offset instead of
 ATH12K_DEFAULT_NOISE_FLOOR

ATH12K_DEFAULT_NOISE_FLOOR is the approximate noise floor value of
the HW, that is used for rssi calculation. Hence use real noise floor
offset value for the calculation of the signal strength. This offset
is saved in ath12k structure (ar->rssi_offsets.rssi_offset) and it is
updated periodically by firmware through wmi event
WMI_PDEV_RSSI_DBM_CONVERSION_PARAMS_INFO_EVENTID.

Signed-off-by: Raj Kumar Bhagat <quic_rajkbhag@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_mon.c | 5 +++--
 drivers/net/wireless/ath/ath12k/dp_tx.c  | 4 ++--
 drivers/net/wireless/ath/ath12k/mac.c    | 5 ++---
 drivers/net/wireless/ath/ath12k/wmi.c    | 2 +-
 4 files changed, 8 insertions(+), 8 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/dp_mon.c
+++ b/drivers/net/wireless/ath/ath12k/dp_mon.c
@@ -2100,7 +2100,7 @@ static void ath12k_dp_mon_update_radiota
 	bool ampdu_present = ppduinfo->userstats[ppduinfo->userid].ampdu_present;
 
 	rxs->flag |= RX_FLAG_MACTIME_START;
-	rxs->signal = ppduinfo->rssi_comb + ATH12K_DEFAULT_NOISE_FLOOR;
+	rxs->signal = ppduinfo->rssi_comb + ar->rssi_offsets.rssi_offset;
 	rxs->nss = ppduinfo->nss + 1;
 
 	if (ampdu_present) {
--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -905,7 +905,7 @@ ath12k_dp_tx_htt_tx_complete_buf(struct
 	if (ts->acked) {
 		if (!(info->flags & IEEE80211_TX_CTL_NO_ACK)) {
 			info->flags |= IEEE80211_TX_STAT_ACK;
-			info->status.ack_signal = ATH12K_DEFAULT_NOISE_FLOOR +
+			info->status.ack_signal = ar->rssi_offsets.rssi_offset +
 						  ts->ack_rssi;
 			info->status.flags = IEEE80211_TX_STATUS_ACK_SIGNAL_VALID;
 		} else {
@@ -1172,7 +1172,7 @@ static void ath12k_dp_tx_complete_msdu(s
 	if (ts.status == HAL_WBM_TQM_REL_REASON_FRAME_ACKED &&
 	    !(info->flags & IEEE80211_TX_CTL_NO_ACK)) {
 		info->flags |= IEEE80211_TX_STAT_ACK;
-		info->status.ack_signal = ATH12K_DEFAULT_NOISE_FLOOR +
+		info->status.ack_signal = ar->rssi_offsets.rssi_offset +
 					  ts.ack_rssi;
 		info->status.flags = IEEE80211_TX_STATUS_ACK_SIGNAL_VALID;
 	}
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -14562,17 +14562,11 @@ static void ath12k_mac_op_sta_statistics
 	sinfo->txrate.flags = arsta->txrate.flags;
 	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_TX_BITRATE);
 
-	/* TODO: Use real NF instead of default one. */
-	sinfo->signal = arsta->rssi_comb + ATH12K_DEFAULT_NOISE_FLOOR;
+	sinfo->signal = arsta->rssi_comb + ar->rssi_offsets.rssi_offset;
 	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_SIGNAL);
 
-	if (ar->chan_noise_floor) {
-		sinfo->signal_avg = ewma_avg_rssi_read(&arsta->avg_rssi) +
-				    ar->chan_noise_floor;
-	} else {
-		sinfo->signal_avg = ewma_avg_rssi_read(&arsta->avg_rssi) +
-				    ATH12K_DEFAULT_NOISE_FLOOR;
-	}
+	sinfo->signal_avg = ewma_avg_rssi_read(&arsta->avg_rssi) +
+			    ar->rssi_offsets.rssi_offset;
 
 	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_SIGNAL_AVG);
 }
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -7864,11 +7864,9 @@ static int ath12k_pull_peer_assoc_conf_e
 }
 
 static void ath12k_wmi_pull_pdev_stats_base(const struct wmi_pdev_stats_base *src,
-					    struct ath12k_fw_stats_pdev *dst,
-					    struct ath12k *ar)
+					    struct ath12k_fw_stats_pdev *dst)
 {
 	dst->ch_noise_floor = src->chan_nf;
-	ar->chan_noise_floor = src->chan_nf;
 	dst->tx_frame_count = src->tx_frame_count;
 	dst->rx_frame_count = src->rx_frame_count;
 	dst->rx_clear_count = src->rx_clear_count;
@@ -8024,7 +8022,7 @@ int ath12k_wmi_pull_fw_stats(struct ath1
 		if (!dst)
 			continue;
 
-		ath12k_wmi_pull_pdev_stats_base(&src->base, dst, ar);
+		ath12k_wmi_pull_pdev_stats_base(&src->base, dst);
 		ath12k_wmi_pull_pdev_stats_tx(&src->tx, dst);
 		ath12k_wmi_pull_pdev_stats_rx(&src->rx, dst);
 		list_add_tail(&dst->list, &stats->pdevs);
@@ -11221,7 +11219,7 @@ static void ath12k_mgmt_rx_event(struct
 		status->freq = ieee80211_channel_to_frequency(rx_ev->channel,
 							      status->band);
 
-	status->signal = rx_ev->snr + ATH12K_DEFAULT_NOISE_FLOOR;
+	status->signal = rx_ev->snr + ar->rssi_offsets.rssi_offset;
 	status->rate_idx = ath12k_mac_bitrate_to_idx(sband, rx_ev->rate / 100);
 
 	hdr = (struct ieee80211_hdr *)skb->data;
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -855,7 +855,6 @@ struct ath12k {
 	u32 txpower_scale;
 	u32 power_scale;
 	u32 chan_tx_pwr;
-	s32 chan_noise_floor;
 	u32 num_stations;
 	u32 max_num_stations;
 	bool monitor_present;
